cvmfs/cache_ram.h

```diff
@@ -0,0 +1,191 @@
+   * @param id The hash key
+   * @returns -EBADFD if fd is not valid
+   */
+  virtual int Readahead(int fd);
+
+  /** Get the amount of space to be allocated for a call to @ref StartTxn */
+  virtual uint16_t SizeOfTxn() { return sizeof(Transaction); }
+
+  /**
+   * Start a new transaction, allocating the required memory and initializing the transaction parameters
+   * @param id The hash key
+   * @param size The total size of the new cache entry
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
+  virtual int StartTxn(const shash::Any &id, uint64_t size, void *txn);
+
+  /**
+   * Set the transaction parameters. At present, only @ref type is used for pinned, volatile, etc.
+   * @param description Unused
+   * @param type The type of the entry, e.g. pinned
+   * @param flags Unused
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
+  virtual void CtrlTxn(const std::string &description,
+                       const ObjectType type,
+                       const int flags,
+                       void *txn);
+
+  /**
+   * Copy the given memory region into the transaction buffer. Copying starts at the transaction's current offset
+   * @param buf The source address
+   * @param size The number of bytes to copy
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
+  virtual int64_t Write(const void *buf, uint64_t size, void *txn);
+
+  /**
+   * Seek to the beginning of the transaction buffer
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
+  virtual int Reset(void *txn);
+
+  /**
+   * Commit a transaction and open the resulting cache entry. This is necessary to avoid a race condition in which the
+   * cache entry is evicted between the calls to CommitTxn and Open.
+   * @param txn A pointer to space allocated for storing the transaction details
+   * @returns A file descriptor to the new cache entry, or a negative error from @ref CommitTxn
+   */
+  virtual int OpenFromTxn(void *txn);
+
+  /**
+   * Free the resources allocated for the pending transaction
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
+  virtual int AbortTxn(void *txn);
+
+   /**
+   * Commit a transaction to the cache. If there is not enough free space in the cache, first try to make room by evicting
+   * volatile entries. If there is still not enough room, try evicting regular entries. If there is *still* not enough
+   * space, give up an return failure. Note that evictions only occur if they will produce enough space for the transaction;
+   * if -ENOSPC is returned, the states of the transaction and cache are unchanged.
+   * @param txn A pointer to space allocated for storing the transaction details
+   * @returns -ENOSPC if the transaction would exceed the size of the cache
+   */
+  virtual int CommitTxn(void *txn);
+
+ private:
+  struct ReadOnlyFd {
```

Comment:
It would be good to explain that the null hash serves as an invalid handle, and perhaps why we don't need to know about the size of the object at this point (I guess out-of-bounds reads either don't happen or they are handled by the kvstore implementation).